Add NamedPipe input operator

sinkingpoint opened the pull request:
**Description:** 

This is a namedpipe input operator, which will read from a named pipe and send the data to the pipeline. It pretty closely mimics the file input operator, but with a few differences.

In particular, named pipes have an interesting property that they receive EOFs when a writer closes the pipe, but that _doesn't_ mean that the pipe is closed. To solve this issue, we crib from existing `tail -f` implementations and use an inotify watcher to detect whenever the pipe receives new data, and then read it using the standard `bufio.Scanner` reader.

**Link to tracking Issue:** #27234

**Testing:**

We add a couple of tests for the new operator. The first tests simply the creation of the named pipe - checking that it's created as a pipe, with the right permissions. The second goes further by inserting logs over several different `Open`s into the pipe, testing that the logs are read, and that the operator is able to handle the named pipe behavior of skipping over EOFs.

**Documentation:**

None, at the moment

/cc @djaglowski

sinkingpoint committed:
Add NamedPipe input operator

This is a namedpipe input operator, which will read from a named pipe
and send the data to the pipeline. It pretty closely mimicks the file input operator,
but with a few differences.

In particular, named pipes have an interesting property that they receive
EOFs when a writer closes the pipe, but that _doesn't_ mean that the pipe
is closed. To solve this issue, we crib from existing `tail -f` implementations
and use an inotify watcher to detect whenever the pipe receives new data, and
then read it using the standard `bufio.Scanner` reader.

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

songy23 reviewed (commented):
Please add a changelog

djaglowski commented on pkg/stanza/operator/input/namedpipe/namedpipe.go:
```diff
@@ -0,0 +1,196 @@
+	return &Input{
+		InputOperator: inputOperator,
+
+		buffer:      make([]byte, DefaultMaxLogSize),
```
I assume this may need to be configurable. Should we do it now?

sinkingpoint replied:
Done

djaglowski commented on pkg/stanza/operator/input/namedpipe/namedpipe.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+func (n *Input) Start(p operator.Persister) error {
+	stat, err := os.Stat(n.path)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to stat named pipe: %w", err)
+	}
+
+	if !os.IsNotExist(err) && stat.Mode()&os.ModeNamedPipe == 0 {
+		return fmt.Errorf("path %s is not a named pipe", n.path)
+	}
+
+	if os.IsNotExist(err) {
+		if err := unix.Mkfifo(n.path, n.permissions); err != nil {
+			return fmt.Errorf("failed to create named pipe: %w", err)
+		}
+	}
+
+	// chmod the named pipe because mkfifo respects the umask which may result
+	// in a named pipe with incorrect permissions.
+	if err := os.Chmod(n.path, os.FileMode(n.permissions)); err != nil {
+		return fmt.Errorf("failed to chmod named pipe: %w", err)
+	}
+
+	watcher, err := NewWatcher(n.path)
+	if err != nil {
+		return fmt.Errorf("failed to create watcher: %w", err)
+	}
+
+	pipe, err := os.OpenFile(n.path, os.O_RDWR, os.ModeNamedPipe)
+	if err != nil {
+		return fmt.Errorf("failed to open named pipe: %w", err)
+	}
```
There's a lot of complexity here. If reasonable, please make sure unit tests cover these failure modes.

sinkingpoint replied:
Done - added a few more test cases for these various if switches. I'm wondering if it makes sense to abstract this logic out into its own thing that we can test in isolation, outside the operator creation process. WDYT?

djaglowski commented on pkg/stanza/operator/input/namedpipe/watcher.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+func newWatcher(path string) (*fsnotify.Watcher, error) {
+	watcher, err := fsnotify.NewWatcher()
```
Do you know if named pipes may be subject to this problems describe [here](github.com/fsnotify/fsnotify)?

sinkingpoint replied:
Not sure that link is working?

djaglowski replied:
Sorry, not sure what happened there. I've fixed the link.

sinkingpoint replied:
Cool. Yeah - that doesn't _really_ apply to us. The worry there is that files get rotated out from under the watcher, however here we assume that the collector will manage the pipe, and leave it undefined as to the behavior if something else fiddles with it

djaglowski reviewed (commented):
Overall looks about right to me.

sinkingpoint committed:
Allow configuring MaxLogSize for namedpipe input

This allows the user to configure the maximum size of a log line, rather
than always using the default.

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint committed:
Add more tests for to cover more paths of named pipe creation

Here we add a couple of tests to cover more cases of named pipe creation.
In particular:

 - A test to check that if a file already exists (i.e. not a named pipe),
   fail the operator creation.
 - A test to check that if a _pipe_ already exists, pass the operator
   construction because we can just use the existing pipe.

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint commented:
Rebased and added licenses

sinkingpoint committed:
Add CHANGELOG entry

Signed-off-by: sinkingpoint <colin@quirl.co.nz>

sinkingpoint commented:
So many checks 😅

sinkingpoint commented:
@djaglowski is there a `make` command I can run that runs all of these? Feels like I'm playing cat and mouse a bit

sinkingpoint committed:
make gotidy

djaglowski commented:
Unfortuantely there's not a catchall because so much is only encoded in github actions. `make all` is pretty thorough but takes quite a while to run.
